File: crates/pattern_runtime/haskell/Pattern/Web.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE OverloadedStrings #-}
module Pattern.Web
  ( Web(..)
  , search
  , fetch
  , fetchReadable
  , fetchRaw
  , fetchContinue
  ) where

import Control.Monad.Freer (Eff, Member, send)
import Data.Text (Text)

-- | Web interaction effect: search the web and fetch page content.
data Web a where
  -- | Search the web. Returns JSON array of {title, url, snippet}.
  WebSearch :: Text -> Maybe Int -> Web Text
  -- | Fetch a URL. Format: Nothing/Just "readable" for extracted text,
  --   Just "raw" for HTML.
  WebFetch :: Text -> Maybe Text -> Web Text
  -- | Continue reading a fetched page from a character offset.
  WebFetchContinue :: Text -> Int -> Maybe Int -> Web Text

-- | Search the web for a query. Returns JSON array of search results.
search :: Member Web effs => Text -> Eff effs Text
search q = send (WebSearch q Nothing)

-- | Fetch a URL and extract readable text content.
fetch :: Member Web effs => Text -> Eff effs Text
fetch url = send (WebFetch url Nothing)

-- | Fetch a URL and extract readable text (explicit).
fetchReadable :: Member Web effs => Text -> Eff effs Text
fetchReadable url = send (WebFetch url (Just "readable"))

-- | Fetch a URL and return raw HTML.
fetchRaw :: Member Web effs => Text -> Eff effs Text
fetchRaw url = send (WebFetch url (Just "raw"))

-- | Continue reading from a previous fetch at a character offset.
fetchContinue :: Member Web effs => Text -> Int -> Eff effs Text
fetchContinue url offset = send (WebFetchContinue url offset Nothing)

File: crates/pattern_runtime/src/sdk/requests/web.rs
// Copyright 2026 Pattern contributors
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, you can obtain one at http://mozilla.org/MPL/2.0/.

//! Mirror of `Pattern.Web` (`haskell/Pattern/Web.hs`).

use tidepool_bridge_derive::FromCore;

/// Rust mirror of the Haskell `Web` GADT.
///
/// - `WebSearch(query, limit)`: search the web using Brave/DDG cascade.
///   Returns JSON array of `{title, url, snippet}` objects.
/// - `WebFetch(url, format)`: fetch URL content. Format: "readable" (default)
///   extracts article text, "raw" returns HTML. Returns text content.
/// - `WebFetchContinue(url, offset, limit)`: continue reading a previously
///   fetched page from the given character offset.
#[derive(Debug, FromCore)]
pub enum WebReq {
    #[core(module = "Pattern.Web", name = "WebSearch")]
    WebSearch(String, Option<i64>),
    #[core(module = "Pattern.Web", name = "WebFetch")]
    WebFetch(String, Option<String>),
    #[core(module = "Pattern.Web", name = "WebFetchContinue")]
    WebFetchContinue(String, i64, Option<i64>),
}
